src/metabolic_ninja/helpers.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) 2018, Novo Nordisk Foundation Center for Biosustainability,
+# Technical University of Denmark.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import logging
+import re
+from operator import itemgetter
+
+
+__all__ = ("identify_exotic_cofactors",)
+
+
+logger = logging.getLogger(__name__)
+atom_pattern = re.compile(r"(?P<atom>[A-Z][a-z]?)(?P<count>[0-9]*)")
+
+
+def count_atoms(formula):
+    """Convert a formula string into a dictionary of counts."""
+    return {
+        m.group("atom"):
+        1 if len(m.group("count")) == 0 else int(m.group("count"))
+        for m in atom_pattern.finditer(formula)
+    }
+
+
+def compute_chemical_linkage_strength(counts_a, counts_d):
+    """
+    Compute the strength of chemical linkage (SCL) as in [1]_.
+
+    Parameters
+    ----------
+    counts_a : dict
+        A mapping between atoms as characters and their counts as parsed from a
+        formula.
+    counts_d : dict
+        A mapping between atoms as characters and their counts as parsed from a
+        formula.
+
+    Returns
+    -------
+    float
+        Return the strength of chemical linkage between compound A and B which
+        is the 'intersection' between their atom counts, normalized by the
+        greater of the total atom counts of either A or B.
+
+    References
+    ----------
+    .. [1] Zhou, Wanding and Luay, Nakhleh.
+       "The Strength of Chemical Linkage as a Criterion for Pruning Metabolic
+       Graphs." Bioinformatics 27, no. 14 (July 15, 2011): 1957–63.
+       https://doi.org/10.1093/bioinformatics/btr271.
+
+    """
+    scl = 0
+    a_total = 0
+    d_total = 0
+    atoms = counts_a.keys() | counts_d.keys()
+    # Ignoring hydrogen atoms is a heuristic from the original publication.
+    if "H" in atoms:
+        atoms.remove("H")
+    for atom in atoms:
+        a_freq = counts_a.get(atom, 0)
+        d_freq = counts_d.get(atom, 0)
+        a_total += a_freq
+        d_total += d_freq
+        scl += min(a_freq, d_freq)
+    return scl / max(a_total, d_total)
+
+
+def identify_exotic_cofactors(pathway, solution, threshold=1E-07):
```

Comment:
Could you add tests for this function too? There is quite a lot of logic which means that tests for `compute_chemical_linkage_strength` and `count_atoms` don't cover much of this.